ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -9685,6 +9697,34 @@ static bool ggml_vk_is_empty(ggml_tensor * node) {
     return ggml_is_empty(node) || node->op == GGML_OP_NONE || node->op == GGML_OP_RESHAPE || node->op == GGML_OP_TRANSPOSE || node->op == GGML_OP_VIEW || node->op == GGML_OP_PERMUTE;
 }
 
+// Returns true if nodes [i, i+1] are fusable RMS_NORM + MUL.
+bool ggml_can_fuse_rms_norm_mul(ggml_backend_vk_context * ctx, ggml_cgraph * cgraph, int i) {
+    ggml_tensor *norm = cgraph->nodes[i];
+
+    if (norm->op != GGML_OP_RMS_NORM || norm->use_count != 1) {
+        return false;
+    }
+    // if norm is a view, some other node might be using the intermediate result
+    // view the view source.
+    if (norm->view_src) {
+        return false;
+    }
+
+    if (i + 1 >= cgraph->n_nodes) {
+        return false;
+    }
+    ggml_tensor *mul = cgraph->nodes[i + 1];
+    if (mul->op != GGML_OP_MUL || mul->src[0] != norm) {
+        return false;
+    }
+
+    // Since norm is the first operand of mul, it must be the same shape
+    GGML_ASSERT(ggml_are_same_shape(mul, norm));
+
+    // XXX TODO: Do we need a way to indicate that the user doesn't need the intermediate result?
```

Comment:
@slaren I'd like to get your thoughts on this. How do we know that the user doesn't want to have all the intermediate results available to them? Seems like they'd need to opt in or out of fusion somehow... GGML_TENSOR_FLAG_OUTPUT seems like almost what we would want, but it's slightly different and doesn't seem to be used much?

Reply:
Precisely because getting details like this one right is not completely obvious, it would be good to have a function in ggml similar to this one that checks if multiple operations can be fused, that all the backends can use to implement operator fusion.

Reply:
Yeah, I can move some of the logic there, but I think eventually some of the implementation details may differ between backends. Hopefully there's a way to have some common logic and some backend-specific logic.

Reply:
OK, I've extracted out some common logic into a helper function.

Reply:
Can you also extract the remaining logic in `ggml_can_fuse_rms_norm_mul` to a generic ggml function that takes a number of ops? Similar to what was proposed in https://github.com/ggml-org/llama.cpp/pull/14158#issuecomment-2997750622. While it may not work in every case, it will work in many cases, and it will avoid duplicating this code in every backend.

Reply:
I will, but I'm skeptical that in the long run there will be a lot of commonality.